NumberFormatException  for es.host  while starting Elasticsearch source connector

Hi
I am getting below error when I start connector. 
Please help on this. Its running but Cant see any desired topic in kafka. 


[2021-05-12 16:30:42,611] INFO ElasticSourceConnectorConfig values: 
	batch.max.rows = 10000
	connection.attempts = 3
	connection.backoff.ms = 10000
	es.host = "myhost.uat.com"
	es.password = "myEspass"
	es.port = "9200"
	es.scheme = http
	es.user = "elastic"
	fieldname_converter = avro
	filters.json_cast = null
	filters.whitelist = null
	incrementing.field.name = "@timestamp_log"
	index.prefix = "my_index-*"
	mode = 
	poll.interval.ms = 5000
	topic.prefix = "myprefix_"
 (com.github.dariobalinzo.ElasticSourceConnectorConfig:347)
[2021-05-12 16:30:42,611] ERROR WorkerConnector{id=elasticsearch-source} Error while starting connector (org.apache.kafka.connect.runtime.WorkerConnector:119)
java.lang.NumberFormatException: For input string: ""9200""
	at java.base/java.lang.NumberFormatException.forInputString(NumberFormatException.java:65)
	at java.base/java.lang.Integer.parseInt(Integer.java:638)
	at java.base/java.lang.Integer.parseInt(Integer.java:770)
	at com.github.dariobalinzo.ElasticSourceConnector.start(ElasticSourceConnector.java:63)
	at org.apache.kafka.connect.runtime.WorkerConnector.doStart(WorkerConnector.java:111)
	at org.apache.kafka.connect.runtime.WorkerConnector.start(WorkerConnector.java:136)
	at org.apache.kafka.connect.runtime.WorkerConnector.transitionTo(WorkerConnector.java:196)
	at org.apache.kafka.connect.runtime.Worker.startConnector(Worker.java:253)
	at org.apache.kafka.connect.runtime.standalone.StandaloneHerder.startConnector(StandaloneHerder.java:293)
	at org.apache.kafka.connect.runtime.standalone.StandaloneHerder.putConnectorConfig(StandaloneHerder.java:209)
	at org.apache.kafka.connect.cli.ConnectStandalone.main(ConnectStandalone.java:115)
[2021-05-12 16:30:42,614] INFO Finished creating connector elasticsearch-source (org.apache.kafka.connect.runtime.Worker:269)
[2021-05-12 16:30:42,615] INFO Skipping update of connector elasticsearch-source since it is not running

After changing it to an integre, it gives me below error.
Elasticsearch instance is running well on given host and port and I can connect using CURL. 

Is there any problem of HTTPS/HTTP?   
My Elasticsearch is runnong over HTTPS.  I am using HTTP scheme in connector config because I didnt find any SSL config options in documentation. 

[2021-05-12 19:30:46,241] ERROR error in searching index names (com.github.dariobalinzo.elastic.ElasticRepository:105)
[2021-05-12 19:30:46,242] ERROR Stopping after connector error (org.apache.kafka.connect.cli.ConnectStandalone:121)
java.lang.RuntimeException: java.net.UnknownHostException: "myhost.uat.com": Name or service not known
	at com.github.dariobalinzo.elastic.ElasticRepository.catIndices(ElasticRepository.java:106)
	at com.github.dariobalinzo.ElasticSourceConnector.taskConfigs(ElasticSourceConnector.java:98)
	at org.apache.kafka.connect.runtime.Worker.connectorTaskConfigs(Worker.java:319)
	at org.apache.kafka.connect.runtime.standalone.StandaloneHerder.recomputeTaskConfigs(StandaloneHerder.java:303)
	at org.apache.kafka.connect.runtime.standalone.StandaloneHerder.updateConnectorTasks(StandaloneHerder.java:329)
	at org.apache.kafka.connect.runtime.standalone.StandaloneHerder.putConnectorConfig(StandaloneHerder.java:214)
	at org.apache.kafka.connect.cli.ConnectStandalone.main(ConnectStandalone.java:115)
Caused by: java.net.UnknownHostException: "myhost.uat.com": Name or service not known

--- a/README.md
+++ b/README.md
@@ -131,6 +131,33 @@ to avoid data losses when paginating (available starting from versions >= 1.4).
 * Type: any
 * Importance: low
 
+
+``es.tls.truststore.location``
+Elastic ssl truststore location
+
+* Type: string
+* Importance: medium
+
+``es.tls.truststore.password``
+Elastic ssl truststore password
+
+* Type: string
+* Default: ""
+* Importance: medium
+
+``es.tls.keystore.location``
+Elasticsearch keystore location
+
+* Type: string
+* Importance: medium
+
+``es.tls.keystore.password``
+Elasticsearch keystore password
+
+* Type: string
+* Default: ""
+* Importance: medium
+
 ``connection.attempts``
 Maximum number of attempts to retrieve a valid Elasticsearch connection.
 

--- a/doc/README.md
+++ b/doc/README.md
@@ -124,12 +124,39 @@ The name of the strictly incrementing field to use to detect new records.
 * Importance: high
 
 ``incrementing.secondary.field.name``
-In case the main incrementing field may have duplicates,
-this secondary field is used as a secondary sort field in order
+In case the main incrementing field may have duplicates, 
+this secondary field is used as a secondary sort field in order 
 to avoid data losses when paginating (available starting from versions >= 1.4).
 
 * Type: any
 * Importance: low
+
+
+``es.tls.truststore.location``
+Elastic ssl truststore location
+
+* Type: string
+* Importance: medium
+
+``es.tls.truststore.password``
+Elastic ssl truststore password
+
+* Type: string
+* Default: ""
+* Importance: medium
+
+``es.tls.keystore.location``
+Elasticsearch keystore location
+
+* Type: string
+* Importance: medium
+
+``es.tls.keystore.password``
+Elasticsearch keystore password
+
+* Type: string
+* Default: ""
+* Importance: medium
 
 ``connection.attempts``
 Maximum number of attempts to retrieve a valid Elasticsearch connection.
@@ -150,6 +177,13 @@ Indices prefix to include in copying.
 
 * Type: string
 * Default: ""
+* Importance: medium
+
+``index.names``
+List of elasticsearch indices: `es1,es2,es3`
+
+* Type: string
+* Default: null
 * Importance: medium
 
 ### Connector Configuration
@@ -176,6 +210,15 @@ Prefix to prepend to index names to generate the name of the Kafka topic to publ
 
 ``filters.whitelist``
 Whitelist filter for extracting a subset of fields from elastic-search json documents. The whitelist filter supports
+nested fields. To provide multiple fields use `;` as separator
+(e.g. `customer;order.qty;order.price`).
+
+* Type: string
+* Importance: medium
+* Default: null
+
+``filters.blacklist``
+Blacklist filter for extracting a subset of fields from elastic-search json documents. The blacklist filter supports
 nested fields. To provide multiple fields use `;` as separator
 (e.g. `customer;order.qty;order.price`).
 

--- a/src/main/java/com/github/dariobalinzo/ElasticSourceConnector.java
+++ b/src/main/java/com/github/dariobalinzo/ElasticSourceConnector.java
@@ -76,8 +76,8 @@ public class ElasticSourceConnector extends SourceConnector {
 
         String truststore = config.getString(ElasticSourceConnectorConfig.ES_TRUSTSTORE_CONF);
         String truststorePass = config.getString(ElasticSourceConnectorConfig.ES_TRUSTSTORE_PWD_CONF);
-        String keystore = config.getString(ElasticSourceConnectorConfig.ES_TRUSTSTORE_CONF);
-        String keystorePass = config.getString(ElasticSourceConnectorConfig.ES_TRUSTSTORE_PWD_CONF);
+        String keystore = config.getString(ElasticSourceConnectorConfig.ES_KEYSTORE_CONF);
+        String keystorePass = config.getString(ElasticSourceConnectorConfig.ES_KEYSTORE_PWD_CONF);
 
         if (truststore != null) {
             connectionBuilder.withTrustStore(truststore, truststorePass);

--- a/src/main/java/com/github/dariobalinzo/ElasticSourceConnector.java
+++ b/src/main/java/com/github/dariobalinzo/ElasticSourceConnector.java
@@ -74,6 +74,19 @@ public class ElasticSourceConnector extends SourceConnector {
                 .withMaxAttempts(maxConnectionAttempts)
                 .withBackoff(connectionRetryBackoff);
 
+        String truststore = config.getString(ElasticSourceConnectorConfig.ES_TRUSTSTORE_CONF);
+        String truststorePass = config.getString(ElasticSourceConnectorConfig.ES_TRUSTSTORE_PWD_CONF);
+        String keystore = config.getString(ElasticSourceConnectorConfig.ES_TRUSTSTORE_CONF);
+        String keystorePass = config.getString(ElasticSourceConnectorConfig.ES_TRUSTSTORE_PWD_CONF);
+
+        if (truststore != null) {
+            connectionBuilder.withTrustStore(truststore, truststorePass);
+        }
+
+        if (keystore != null) {
+            connectionBuilder.withKeyStore(keystore, keystorePass);
+        }
+
         if (esUser == null || esUser.isEmpty()) {
             elasticConnection = connectionBuilder.build();
         } else {
@@ -81,6 +94,7 @@ public class ElasticSourceConnector extends SourceConnector {
                     .withPassword(esPwd)
                     .build();
         }
+
         elasticRepository = new ElasticRepository(elasticConnection);
     }
 

--- a/src/main/java/com/github/dariobalinzo/ElasticSourceConnectorConfig.java
+++ b/src/main/java/com/github/dariobalinzo/ElasticSourceConnectorConfig.java
@@ -50,6 +50,18 @@ public class ElasticSourceConnectorConfig extends AbstractConfig {
     private final static String ES_PWD_DOC = "Elasticsearch password";
     private final static String ES_PWD_DISPLAY = "Elasticsearch password";
 
+    public final static String ES_KEYSTORE_CONF = "es.tls.keystore.location";
+    private final static String ES_KEYSTORE_DOC = "Elasticsearch keystore location";
+
+    public final static String ES_KEYSTORE_PWD_CONF = "es.tls.keystore.password";
+    private final static String ES_KEYSTORE_PWD_DOC = "Elasticsearch keystore password";
+
+    public final static String ES_TRUSTSTORE_CONF = "es.tls.truststore.location";
+    private final static String ES_TRUSTSTORE_DOC = "Elasticsearch truststore location";
+
+    public final static String ES_TRUSTSTORE_PWD_CONF = "es.tls.keystore.location";
+    private final static String ES_TRUSTSTORE_PWD_DOC = "Elasticsearch truststore password";
+
     public static final String CONNECTION_ATTEMPTS_CONFIG = "connection.attempts";
     private static final String CONNECTION_ATTEMPTS_DOC
             = "Maximum number of attempts to retrieve a valid Elasticsearch connection.";
@@ -200,6 +212,46 @@ public class ElasticSourceConnectorConfig extends AbstractConfig {
                 ++orderInGroup,
                 Width.SHORT,
                 ES_PWD_DISPLAY
+        ).define(
+                ES_KEYSTORE_CONF,
+                Type.STRING,
+                null,
+                Importance.MEDIUM,
+                ES_KEYSTORE_DOC,
+                DATABASE_GROUP,
+                ++orderInGroup,
+                Width.SHORT,
+                ES_KEYSTORE_DOC
+        ).define(
+                ES_KEYSTORE_PWD_CONF,
+                Type.STRING,
+                null,
+                Importance.MEDIUM,
+                ES_KEYSTORE_PWD_DOC,
+                DATABASE_GROUP,
+                ++orderInGroup,
+                Width.SHORT,
+                ES_KEYSTORE_PWD_DOC
+        ).define(
+                ES_TRUSTSTORE_CONF,
+                Type.STRING,
+                null,
+                Importance.MEDIUM,
+                ES_TRUSTSTORE_DOC,
+                DATABASE_GROUP,
+                ++orderInGroup,
+                Width.SHORT,
+                ES_TRUSTSTORE_DOC
+        ).define(
+                ES_TRUSTSTORE_PWD_CONF,
+                Type.STRING,
+                null,
+                Importance.MEDIUM,
+                ES_TRUSTSTORE_PWD_DOC,
+                DATABASE_GROUP,
+                ++orderInGroup,
+                Width.SHORT,
+                ES_TRUSTSTORE_PWD_DOC
         ).define(
                 CONNECTION_ATTEMPTS_CONFIG,
                 Type.STRING,

--- a/src/main/java/com/github/dariobalinzo/ElasticSourceConnectorConfig.java
+++ b/src/main/java/com/github/dariobalinzo/ElasticSourceConnectorConfig.java
@@ -59,7 +59,7 @@ public class ElasticSourceConnectorConfig extends AbstractConfig {
     public final static String ES_TRUSTSTORE_CONF = "es.tls.truststore.location";
     private final static String ES_TRUSTSTORE_DOC = "Elasticsearch truststore location";
 
-    public final static String ES_TRUSTSTORE_PWD_CONF = "es.tls.keystore.location";
+    public final static String ES_TRUSTSTORE_PWD_CONF = "es.tls.truststore.password";
     private final static String ES_TRUSTSTORE_PWD_DOC = "Elasticsearch truststore password";
 
     public static final String CONNECTION_ATTEMPTS_CONFIG = "connection.attempts";
@@ -225,7 +225,7 @@ public class ElasticSourceConnectorConfig extends AbstractConfig {
         ).define(
                 ES_KEYSTORE_PWD_CONF,
                 Type.STRING,
-                null,
+                "",
                 Importance.MEDIUM,
                 ES_KEYSTORE_PWD_DOC,
                 DATABASE_GROUP,
@@ -245,7 +245,7 @@ public class ElasticSourceConnectorConfig extends AbstractConfig {
         ).define(
                 ES_TRUSTSTORE_PWD_CONF,
                 Type.STRING,
-                null,
+                "",
                 Importance.MEDIUM,
                 ES_TRUSTSTORE_PWD_DOC,
                 DATABASE_GROUP,

--- a/src/main/java/com/github/dariobalinzo/elastic/ElasticConnection.java
+++ b/src/main/java/com/github/dariobalinzo/elastic/ElasticConnection.java
@@ -21,13 +21,22 @@ import org.apache.http.auth.AuthScope;
 import org.apache.http.auth.UsernamePasswordCredentials;
 import org.apache.http.client.CredentialsProvider;
 import org.apache.http.impl.client.BasicCredentialsProvider;
+import org.apache.http.ssl.SSLContextBuilder;
+import org.apache.http.ssl.SSLContexts;
 import org.elasticsearch.client.RestClient;
 import org.elasticsearch.client.RestHighLevelClient;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import javax.net.ssl.SSLContext;
 import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.security.KeyStore;
 import java.util.Arrays;
+import java.util.Objects;
 
 public class ElasticConnection {
     public final static Logger logger = LoggerFactory.getLogger(ElasticConnection.class);
@@ -35,45 +44,89 @@ public class ElasticConnection {
     private RestHighLevelClient client;
     private final long connectionRetryBackoff;
     private final int maxConnectionAttempts;
+    private final String hosts;
+    private final String protocol;
+    private final int port;
+    private final SSLContext sslContext;
+    private final CredentialsProvider credentialsProvider;
 
     ElasticConnection(ElasticConnectionBuilder builder) {
-        String hosts = builder.hosts;
+        hosts = builder.hosts;
+        protocol = builder.protocol;
+        port = builder.port;
+
         String user = builder.user;
         String pwd = builder.pwd;
-        String protocol = builder.protocol;
-        int port = builder.port;
-
-        if (user == null) {
-            createConnectionWithNoAuth(hosts, protocol, port);
+        if (user != null) {
+            credentialsProvider = new BasicCredentialsProvider();
+            credentialsProvider.setCredentials(AuthScope.ANY,
+                    new UsernamePasswordCredentials(user, pwd));
         } else {
-            createConnectionUsingAuth(hosts, protocol, port, user, pwd);
+            credentialsProvider = null;
         }
+
+        sslContext = builder.trustStorePath == null ? null :
+                getSslContext(
+                    builder.trustStorePath,
+                    builder.trustStorePassword,
+                    builder.keyStorePath,
+                    builder.keyStorePassword
+                );
+
+        createConnection();
 
         this.maxConnectionAttempts = builder.maxConnectionAttempts;
         this.connectionRetryBackoff = builder.connectionRetryBackoff;
     }
 
-    private void createConnectionWithNoAuth(String hosts, String protocol, int port) {
-        logger.info("elastic auth disabled");
-        HttpHost[] hostList = parseHosts(hosts, protocol, port);
-        client = new RestHighLevelClient(RestClient.builder(hostList));
-    }
-
-    private void createConnectionUsingAuth(String hosts, String protocol, int port, String user, String pwd) {
-        logger.info("elastic auth enabled");
-        CredentialsProvider credentialsProvider = new BasicCredentialsProvider();
-        credentialsProvider.setCredentials(AuthScope.ANY,
-                new UsernamePasswordCredentials(user, pwd));
-
-
+    private void createConnection() {
         HttpHost[] hostList = parseHosts(hosts, protocol, port);
 
         client = new RestHighLevelClient(
                 RestClient.builder(hostList)
                         .setHttpClientConfigCallback(
-                                httpClientBuilder -> httpClientBuilder.setDefaultCredentialsProvider(credentialsProvider)
+                                httpClientBuilder -> {
+                                    if (credentialsProvider != null) {
+                                        httpClientBuilder.setDefaultCredentialsProvider(credentialsProvider);
+                                    }
+                                    if (sslContext != null) {
+                                        httpClientBuilder.setSSLContext(sslContext);
+                                    }
+                                    return httpClientBuilder;
+                                }
                         )
         );
+    }
+
+    private SSLContext getSslContext(String trustStoreConf, String trustStorePass,
+                             String keyStoreConf, String keyStorePass) {
+
+        Objects.requireNonNull(trustStoreConf, "truststore location is required");
+        Objects.requireNonNull(trustStorePass, "truststore password is required");
+
+        try {
+            Path trustStorePath = Paths.get(trustStoreConf);
+            KeyStore truststore = KeyStore.getInstance("pkcs12");
+            try (InputStream is = Files.newInputStream(trustStorePath)) {
+                truststore.load(is, trustStorePass.toCharArray());
+            }
+            SSLContextBuilder sslBuilder = SSLContexts.custom()
+                    .loadTrustMaterial(truststore, null);
+
+            if (keyStoreConf != null) {
+                Objects.requireNonNull(keyStorePass, "keystore password is required");
+                Path keyStorePath = Paths.get(keyStoreConf);
+                KeyStore keyStore = KeyStore.getInstance("pkcs12");
+                try (InputStream is = Files.newInputStream(keyStorePath)) {
+                    keyStore.load(is, keyStorePass.toCharArray());
+                }
+                sslBuilder.loadKeyMaterial(keyStore, keyStorePass.toCharArray());
+            }
+
+            return sslBuilder.build();
+        } catch (Exception e) {
+            throw new SslContextException(e);
+        }
     }
 
     private HttpHost[] parseHosts(String hosts, String protocol, int port) {

--- a/src/main/java/com/github/dariobalinzo/elastic/ElasticConnectionBuilder.java
+++ b/src/main/java/com/github/dariobalinzo/elastic/ElasticConnectionBuilder.java
@@ -26,6 +26,11 @@ public class ElasticConnectionBuilder {
     String user;
     String pwd;
 
+    String trustStorePath;
+    String trustStorePassword;
+    String keyStorePath;
+    String keyStorePassword;
+
     public ElasticConnectionBuilder(String hosts, int port) {
         this.hosts = hosts;
         this.port = port;
@@ -53,6 +58,18 @@ public class ElasticConnectionBuilder {
 
     public ElasticConnectionBuilder withBackoff(long connectionRetryBackoff) {
         this.connectionRetryBackoff = connectionRetryBackoff;
+        return this;
+    }
+
+    public ElasticConnectionBuilder withTrustStore(String path, String password) {
+        this.trustStorePath = path;
+        this.trustStorePassword = password;
+        return this;
+    }
+
+    public ElasticConnectionBuilder withKeyStore(String path, String password) {
+        this.keyStorePath = path;
+        this.keyStorePassword = password;
         return this;
     }
 

--- a/src/main/java/com/github/dariobalinzo/elastic/SslContextException.java
+++ b/src/main/java/com/github/dariobalinzo/elastic/SslContextException.java
@@ -1,0 +1,7 @@
+package com.github.dariobalinzo.elastic;
+
+public class SslContextException extends RuntimeException {
+    public SslContextException(Exception e) {
+        super(e);
+    }
+}

--- a/src/main/java/com/github/dariobalinzo/task/ElasticSourceTask.java
+++ b/src/main/java/com/github/dariobalinzo/task/ElasticSourceTask.java
@@ -163,6 +163,19 @@ public class ElasticSourceTask extends SourceTask {
                 .withMaxAttempts(maxConnectionAttempts)
                 .withBackoff(connectionRetryBackoff);
 
+        String truststore = config.getString(ElasticSourceConnectorConfig.ES_TRUSTSTORE_CONF);
+        String truststorePass = config.getString(ElasticSourceConnectorConfig.ES_TRUSTSTORE_PWD_CONF);
+        String keystore = config.getString(ElasticSourceConnectorConfig.ES_TRUSTSTORE_CONF);
+        String keystorePass = config.getString(ElasticSourceConnectorConfig.ES_TRUSTSTORE_PWD_CONF);
+
+        if (truststore != null) {
+            connectionBuilder.withTrustStore(truststore, truststorePass);
+        }
+
+        if (keystore != null) {
+            connectionBuilder.withKeyStore(keystore, keystorePass);
+        }
+
         if (esUser == null || esUser.isEmpty()) {
             es = connectionBuilder.build();
         } else {

--- a/src/main/java/com/github/dariobalinzo/task/ElasticSourceTask.java
+++ b/src/main/java/com/github/dariobalinzo/task/ElasticSourceTask.java
@@ -92,6 +92,8 @@ public class ElasticSourceTask extends SourceTask {
 
         topic = config.getString(ElasticSourceConnectorConfig.TOPIC_PREFIX_CONFIG);
         cursorField = config.getString(ElasticSourceConnectorConfig.INCREMENTING_FIELD_NAME_CONFIG);
+        Objects.requireNonNull(cursorField, ElasticSourceConnectorConfig.INCREMENTING_FIELD_NAME_CONFIG
+                + " conf is mandatory");
         cursorFieldJsonName = removeKeywordSuffix(cursorField);
         secondaryCursorField = config.getString(ElasticSourceConnectorConfig.SECONDARY_INCREMENTING_FIELD_NAME_CONFIG);
         secondaryCursorFieldJsonName = removeKeywordSuffix(secondaryCursorField);
@@ -165,8 +167,8 @@ public class ElasticSourceTask extends SourceTask {
 
         String truststore = config.getString(ElasticSourceConnectorConfig.ES_TRUSTSTORE_CONF);
         String truststorePass = config.getString(ElasticSourceConnectorConfig.ES_TRUSTSTORE_PWD_CONF);
-        String keystore = config.getString(ElasticSourceConnectorConfig.ES_TRUSTSTORE_CONF);
-        String keystorePass = config.getString(ElasticSourceConnectorConfig.ES_TRUSTSTORE_PWD_CONF);
+        String keystore = config.getString(ElasticSourceConnectorConfig.ES_KEYSTORE_CONF);
+        String keystorePass = config.getString(ElasticSourceConnectorConfig.ES_KEYSTORE_PWD_CONF);
 
         if (truststore != null) {
             connectionBuilder.withTrustStore(truststore, truststorePass);

